openmeter/billing/adapter/profile.go

```diff
@@ -0,0 +1,319 @@
+	"github.com/openmeterio/openmeter/openmeter/ent/db/billingprofile"
+	"github.com/openmeterio/openmeter/openmeter/ent/db/billingworkflowconfig"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+var _ billing.ProfileAdapter = (*adapter)(nil)
+
+func (a adapter) CreateProfile(ctx context.Context, input billing.CreateProfileInput) (*billing.Profile, error) {
+	if a.tx == nil {
+		return nil, fmt.Errorf("cannot create profile: %w", ErrTransactionRequired)
+	}
+
+	c := a.client()
+
+	dbWorkflowConfig, err := c.BillingWorkflowConfig.Create().
+		SetNamespace(input.Namespace).
+		SetNillableTimezone(input.WorkflowConfig.Timezone).
+		SetCollectionAlignment(input.WorkflowConfig.Collection.Alignment).
+		SetItemCollectionPeriodSeconds(int64(input.WorkflowConfig.Collection.ItemCollectionPeriod / time.Second)).
+		SetInvoiceAutoAdvance(input.WorkflowConfig.Invoicing.AutoAdvance).
+		SetInvoiceDraftPeriodSeconds(int64(input.WorkflowConfig.Invoicing.DraftPeriod / time.Second)).
+		SetInvoiceDueAfterSeconds(int64(input.WorkflowConfig.Invoicing.DueAfter / time.Second)).
+		SetInvoiceItemResolution(input.WorkflowConfig.Invoicing.ItemResolution).
+		SetInvoiceItemPerSubject(input.WorkflowConfig.Invoicing.ItemPerSubject).
+		SetInvoiceCollectionMethod(input.WorkflowConfig.Payment.CollectionMethod).
+		Save(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dbProfile, err := c.BillingProfile.Create().
+		SetNamespace(input.Namespace).
+		SetKey(input.Key).
+		SetDefault(input.Default).
+		SetTaxProvider(input.TaxConfiguration.Type).
+		SetInvoicingProvider(input.InvoicingConfiguration.Type).
+		SetPaymentProvider(input.PaymentConfiguration.Type).
+		SetSupplierName(input.Supplier.Name).
+		SetSupplierAddressCountry(*input.Supplier.Address.Country). // Validation is done at service level
+		SetNillableSupplierAddressState(input.Supplier.Address.State).
+		SetNillableSupplierAddressCity(input.Supplier.Address.City).
+		SetNillableSupplierAddressPostalCode(input.Supplier.Address.PostalCode).
+		SetNillableSupplierAddressLine1(input.Supplier.Address.Line1).
+		SetNillableSupplierAddressLine2(input.Supplier.Address.Line2).
+		SetNillableSupplierAddressPhoneNumber(input.Supplier.Address.PhoneNumber).
+		SetWorkflowConfig(dbWorkflowConfig).
+		Save(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hack: we need to add the edges back
+	dbProfile.Edges.WorkflowConfig = dbWorkflowConfig
+
+	return mapProfileFromDB(dbProfile), nil
+}
+
+func (a adapter) GetProfileByKeyOrID(ctx context.Context, input billing.GetProfileByKeyOrIDInput) (*billing.Profile, error) {
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	dbProfiles, err := a.client().BillingProfile.Query().
+		Where(billingprofile.Namespace(input.Namespace)).
+		Where(billingprofile.Or(
+			billingprofile.And(billingprofile.Key(input.IDOrKey), billingprofile.DeletedAtIsNil()),
+			billingprofile.ID(input.IDOrKey),
+		)).
+		WithWorkflowConfig().All(ctx)
```

Comment:
You can do `query.First(ctx)`

Reply:
No, in case of a key ID collusion this prioritizes keys.